### DISCUSSION TITLE
Why does C# call a 'less better' overload?

### DISCUSSION THREAD

**Discussion by AArnott** [dotnet/csharplang#6003] [Q&A] [answered]
Consider this example:

```cs
using System;

ReadOnlyMemory<char> memory = new Char[] { 'a' };
Console.WriteLine("a".Equals(memory));
Console.WriteLine(Extensions.Equals("a", memory));

static class Extensions{
    internal static bool Equals(this string value, ReadOnlyMemory<char> memory) => true;
}
```

This produces "False, True". Why? Because the C# compiler binds the `"a".Equals` method call to `string.Equals(object?)` instead of the `Extensions.Equals(this string, ReadOnlyMemory<char>)` extension method. Why? The extension method clearly matches the type better.

[Sharplab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQBMIGMD2A7DCMKOAGxgBoYQE48L0QBqAHwAEAmARgFgAofgCVo6APJ5SATwCyEALY4EkgDxYAFlAQA+AARz5iyToC8OvBADuOgMIaEAbQC6OgN46A5FHc6AvgG5+Ni4ATgAKACIocIA6AFEARzgoUmRQ/QUlAEpMgL4gsNiADxgIPGQAS3xkOMTk1Mjwij0DLJz+QK4ANh1OHSKSssqyl34dMZ1yvBKEPGSerp1gHBxSPtqU0Jg1cuR5gAYdADdkuAgm4SgxCRkWlXVNXXTDTJNdGFoIXJ8gA===)

**Comment by HaloFour** [comment, marked as answer]
Extensions aren't overloads, they're only considered if there are no instance members that match, even if they require type coercion.

**Comment by bernd5** [comment]
[See: 11.7.8.3 Extension method invocations](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/expressions.md#11783-extension-method-invocations)

> if the normal processing of the invocation finds no applicable methods, an attempt is made to process the construct as an extension method invocation. If «expr» or any of the «args» has compile-time type dynamic, extension methods will not apply.

**Reply by AArnott** [reply]
Thank you for your spec-based answer.
